### PR TITLE
JBIDE-21371 - add custom auto-publish thread    

### PR DESCRIPTION
**Pull request by robstryker**
and ensure publishing for special project happens
   automatically

### CODE REVIEW

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/resources/OpenshiftResourceChangeListener.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
+package org.jboss.tools.openshift.core.server.resources;
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/resources/OpenshiftResourceChangeListener.java]: missing copyright header

Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/resources/OpenshiftResourceChangeListener.java]: should be internal package

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/NewApplicationHandler.java [hunk 2]
```diff
@@ -32,6 +32,10 @@
 	public Object execute(ExecutionEvent event) throws ExecutionException {
 		try {
 			WizardUtils.openWizardDialog(new NewApplicationWizard(), HandlerUtil.getActiveShell(event));
+
+			// This technically sends the wrong event for the wrong 'command' id,
+			// but the refresh resource handler only looks at the selection... 
+			new RefreshResourceHandler().execute(event);
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/NewApplicationHandler.java]: how is that related to the autopublish issue?

Comment (robstryker) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/NewApplicationHandler.java]: It's not, but I just got tired of the viewer not being updated when i exited the wizard. plz don't make me open new issue :(

Comment (adietish) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/NewApplicationHandler.java]: This workaround will only refresh the selected resource, so this might help you when coding but I dont think that this is even a valid general purpose workaround. I would remove it.

Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/handler/NewApplicationHandler.java]: I believe the proper fix for refresh will be available from the new openshift explorer layout feature branch, so this hack should be removed

**Comment by fbricon** [comment]
+1 to merge after removing new RefreshResourceHandler().execute(event); and squashing the commits

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/resources/OpenshiftResourceChangeListener.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+	public class MagicProjectChangeJob extends Job {
+		private IResourceDelta delta;
+		private OpenShiftServer openshiftServer;
+		private IResourceChangeEvent event;
+		private List<IResource> changes;
+
+		public MagicProjectChangeJob(OpenShiftServer openshiftServer, IResourceDelta delta, List<IResource> changes) {
+			super(NLS.bind(Messages.jobUpdateServer, openshiftServer.getServer().getName()));
+			this.openshiftServer = openshiftServer;
+			this.delta = delta;
```
Comment (scabanovich) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/resources/OpenshiftResourceChangeListener.java]: IResourceDelta passed to MagicProjectChangeJob is never used which is good because as docs warn the resource delta object 'is valid only for the duration of the invocation of' resourceChanged(IResourceChangeEvent), it may not be valid when job is runned. I suggest removing this parameter.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/resources/OpenshiftResourceChangeListener.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+			((Server)server).setServerPublishState(newState);
+			
+			// TODO set the delta?
+			
+			
+			if (server.getServerState() != IServer.STATE_STOPPED && behaviourDelegate != null)
+				behaviourDelegate.handleResourceChange();
+			
+			if (server.getServerState() == IServer.STATE_STARTED)
+				autoPublish(event);
```
Comment (scabanovich) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/resources/OpenshiftResourceChangeListener.java]: Object event was not passed to MagicProjectChangeJob.
